src/functional/fencil_processors/pipeline.py

```diff
@@ -0,0 +1,40 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from typing import Callable, Optional, Protocol, TypeVar
+
+from .builders.cache import Strategy as CacheStrategy
+from .source_modules.source_modules import BindingModule, SourceModule, SupportedLanguage
+
+
+LanguageT_contra = TypeVar("LanguageT_contra", bound=SupportedLanguage, contravariant=True)
+
+
+class BindingsGenerator(Protocol[LanguageT_contra]):
+    def __call__(self, source_module: SourceModule[LanguageT_contra]) -> BindingModule:
```

Comment:
Just a question: why is the type not covariantly parametrized with respect to the output binding language?

Reply:
This took me a while to get (mypy complained about using the covariant type var), because I hadn't understood co- and contravariant all the way.

A BindingsGenerator written for the most general language protocol works with specific ones but not vice-versa. So you can substitute *super*-classes for *sub*-classes (contravariant) and not *sub*-classes for *super*-classes (covariant).

Reply:
On the other hand, now mypy complains the other way around, so I apparently don't understand anything.

Reply:
In my experience, understanding covariance and contravariance is much simpler than understanding what the hell is mypy understanding for covariance and contravariance. I almost gave up completely on using covariant and contravariant typevars in Python annotations...

Reply:
side note: what mypy expects depends on the presence (or not) of the infamous `src/__init__.py`. So this might need to be revisited...